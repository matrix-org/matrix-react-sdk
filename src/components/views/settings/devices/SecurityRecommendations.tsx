/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from "react";

import { _t } from "../../../../languageHandler";
import AccessibleButton from "../../elements/AccessibleButton";
import SettingsSubsection from "../shared/SettingsSubsection";
import DeviceSecurityCard from "./DeviceSecurityCard";
import { DeviceSecurityLearnMore } from "./DeviceSecurityLearnMore";
import { filterDevicesBySecurityRecommendation, FilterVariation, INACTIVE_DEVICE_AGE_DAYS } from "./filter";
import { DeviceSecurityVariation, ExtendedDevice, DevicesDictionary } from "./types";

interface Props {
    devices: DevicesDictionary;
    currentDeviceId: ExtendedDevice["device_id"];
    goToFilteredList: (filter: FilterVariation) => void;
}

const SecurityRecommendations: React.FC<Props> = ({ devices, currentDeviceId, goToFilteredList }) => {
    const devicesArray = Object.values<ExtendedDevice>(devices);

    const unverifiedDevicesCount = filterDevicesBySecurityRecommendation(devicesArray, [
        DeviceSecurityVariation.Unverified,
    ])
        // filter out the current device
        // as unverfied warning and actions
        // will be shown in current session section
        .filter((device) => device.device_id !== currentDeviceId).length;
    const inactiveDevicesCount = filterDevicesBySecurityRecommendation(devicesArray, [
        DeviceSecurityVariation.Inactive,
    ]).length;

    if (!(unverifiedDevicesCount | inactiveDevicesCount)) {
        return null;
    }

    const inactiveAgeDays = INACTIVE_DEVICE_AGE_DAYS;

    return (
        <SettingsSubsection
            heading={_t("settings|sessions|security_recommendations")}
            description={_t("settings|sessions|security_recommendations_description")}
            data-testid="security-recommendations-section"
        >
            {!!unverifiedDevicesCount && (
                <DeviceSecurityCard
                    variation={DeviceSecurityVariation.Unverified}
                    heading={_t("settings|sessions|unverified_sessions")}
                    description={
                        <>
                            {_t("settings|sessions|unverified_sessions_list_description")}
                            <DeviceSecurityLearnMore variation={DeviceSecurityVariation.Unverified} />
                        </>
                    }
                >
                    <AccessibleButton
                        kind="link_inline"
                        onClick={() => goToFilteredList(DeviceSecurityVariation.Unverified)}
                        data-testid="unverified-devices-cta"
                    >
                        {_t("action|view_all") + ` (${unverifiedDevicesCount})`}
                    </AccessibleButton>
                </DeviceSecurityCard>
            )}
            {!!inactiveDevicesCount && (
                <>
                    {!!unverifiedDevicesCount && <div className="mx_SecurityRecommendations_spacing" />}
                    <DeviceSecurityCard
                        variation={DeviceSecurityVariation.Inactive}
                        heading={_t("settings|sessions|inactive_sessions")}
                        description={
                            <>
                                {_t("settings|sessions|inactive_sessions_list_description", { inactiveAgeDays })}
                                <DeviceSecurityLearnMore variation={DeviceSecurityVariation.Inactive} />
                            </>
                        }
                    >
                        <AccessibleButton
                            kind="link_inline"
                            onClick={() => goToFilteredList(DeviceSecurityVariation.Inactive)}
                            data-testid="inactive-devices-cta"
                        >
                            {_t("action|view_all") + ` (${inactiveDevicesCount})`}
                        </AccessibleButton>
                    </DeviceSecurityCard>
                </>
            )}
        </SettingsSubsection>
    );
};

export default SecurityRecommendations;
