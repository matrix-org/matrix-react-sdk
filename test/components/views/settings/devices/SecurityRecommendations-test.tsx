
import React from 'react';
import { mount } from 'enzyme';

import SecurityRecommendations from '../../../../../src/components/views/settings/devices/SecurityRecommendations';

describe('<SecurityRecommendations />', () => {
    const defaultProps = {};
    const getComponent = (props = {}) =>
        mount(<SecurityRecommendations {...defaultProps} {...props} />);

    it('renders', () => {
        const component = getComponent();
        expect(component).toBeTruthy();
    });
});
