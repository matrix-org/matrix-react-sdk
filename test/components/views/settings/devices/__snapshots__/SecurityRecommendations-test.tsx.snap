// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SecurityRecommendations /> renders both cards when user has both unverified and inactive devices 1`] = `
<div>
  <div
    class="mx_SettingsSubsection"
    data-testid="security-recommendations-section"
  >
    <h3
      class="mx_Heading_h3 mx_SettingsSubsection_heading"
    >
      Security recommendations
    </h3>
    <div
      class="mx_SettingsSubsection_description"
    >
      Improve your account security by following these recommendations
    </div>
    <div
      class="mx_SettingsSubsection_content"
    >
      <div
        class="mx_DeviceSecurityCard"
      >
        <div
          class="mx_DeviceSecurityCard_icon Unverified"
        >
          <div
            height="16"
            width="16"
          />
        </div>
        <div
          class="mx_DeviceSecurityCard_content"
        >
          <p
            class="mx_DeviceSecurityCard_heading"
          >
            Unverified sessions
          </p>
          <p
            class="mx_DeviceSecurityCard_description"
          >
            Verify your sessions for enhanced secure messaging or sign out from those you don't recognize or use anymore.
          </p>
          <div
            class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_link_inline"
            role="button"
            tabindex="0"
          >
            View all (1)
          </div>
        </div>
      </div>
      <div
        class="mx_SecurityRecommendations_spacing"
      />
      <div
        class="mx_DeviceSecurityCard"
      >
        <div
          class="mx_DeviceSecurityCard_icon Inactive"
        >
          <div
            height="16"
            width="16"
          />
        </div>
        <div
          class="mx_DeviceSecurityCard_content"
        >
          <p
            class="mx_DeviceSecurityCard_heading"
          >
            Inactive sessions
          </p>
          <p
            class="mx_DeviceSecurityCard_description"
          >
            Consider signing out from old sessions (90 days or older) you don't use anymore
          </p>
          <div
            class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_link_inline"
            role="button"
            tabindex="0"
          >
            View all (1)
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<SecurityRecommendations /> renders inactive devices section when user has inactive devices 1`] = `
<div>
  <div
    class="mx_SettingsSubsection"
    data-testid="security-recommendations-section"
  >
    <h3
      class="mx_Heading_h3 mx_SettingsSubsection_heading"
    >
      Security recommendations
    </h3>
    <div
      class="mx_SettingsSubsection_description"
    >
      Improve your account security by following these recommendations
    </div>
    <div
      class="mx_SettingsSubsection_content"
    >
      <div
        class="mx_DeviceSecurityCard"
      >
        <div
          class="mx_DeviceSecurityCard_icon Unverified"
        >
          <div
            height="16"
            width="16"
          />
        </div>
        <div
          class="mx_DeviceSecurityCard_content"
        >
          <p
            class="mx_DeviceSecurityCard_heading"
          >
            Unverified sessions
          </p>
          <p
            class="mx_DeviceSecurityCard_description"
          >
            Verify your sessions for enhanced secure messaging or sign out from those you don't recognize or use anymore.
          </p>
          <div
            class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_link_inline"
            role="button"
            tabindex="0"
          >
            View all (1)
          </div>
        </div>
      </div>
      <div
        class="mx_SecurityRecommendations_spacing"
      />
      <div
        class="mx_DeviceSecurityCard"
      >
        <div
          class="mx_DeviceSecurityCard_icon Inactive"
        >
          <div
            height="16"
            width="16"
          />
        </div>
        <div
          class="mx_DeviceSecurityCard_content"
        >
          <p
            class="mx_DeviceSecurityCard_heading"
          >
            Inactive sessions
          </p>
          <p
            class="mx_DeviceSecurityCard_description"
          >
            Consider signing out from old sessions (90 days or older) you don't use anymore
          </p>
          <div
            class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_link_inline"
            role="button"
            tabindex="0"
          >
            View all (1)
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<SecurityRecommendations /> renders unverified devices section when user has unverified devices 1`] = `
<div>
  <div
    class="mx_SettingsSubsection"
    data-testid="security-recommendations-section"
  >
    <h3
      class="mx_Heading_h3 mx_SettingsSubsection_heading"
    >
      Security recommendations
    </h3>
    <div
      class="mx_SettingsSubsection_description"
    >
      Improve your account security by following these recommendations
    </div>
    <div
      class="mx_SettingsSubsection_content"
    >
      <div
        class="mx_DeviceSecurityCard"
      >
        <div
          class="mx_DeviceSecurityCard_icon Unverified"
        >
          <div
            height="16"
            width="16"
          />
        </div>
        <div
          class="mx_DeviceSecurityCard_content"
        >
          <p
            class="mx_DeviceSecurityCard_heading"
          >
            Unverified sessions
          </p>
          <p
            class="mx_DeviceSecurityCard_description"
          >
            Verify your sessions for enhanced secure messaging or sign out from those you don't recognize or use anymore.
          </p>
          <div
            class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_link_inline"
            role="button"
            tabindex="0"
          >
            View all (2)
          </div>
        </div>
      </div>
      <div
        class="mx_SecurityRecommendations_spacing"
      />
      <div
        class="mx_DeviceSecurityCard"
      >
        <div
          class="mx_DeviceSecurityCard_icon Inactive"
        >
          <div
            height="16"
            width="16"
          />
        </div>
        <div
          class="mx_DeviceSecurityCard_content"
        >
          <p
            class="mx_DeviceSecurityCard_heading"
          >
            Inactive sessions
          </p>
          <p
            class="mx_DeviceSecurityCard_description"
          >
            Consider signing out from old sessions (90 days or older) you don't use anymore
          </p>
          <div
            class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_link_inline"
            role="button"
            tabindex="0"
          >
            View all (1)
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
